articles/working-with-pythons-unittest/unittests.md

```diff
@@ -0,0 +1,221 @@
+FAIL: test_double (tests.TestMathFunctions)
+----------------------------------------------------------------------
+Traceback (most recent call last):
+  File "main.py", line 8, in test_double
+    self.assertEqual(double(8), 16)
+AssertionError: 32 != 16
+
+----------------------------------------------------------------------
+Ran 3 tests in 0.001s
+
+FAILED (failures=1)
+```
+
+##### Error messages
+Error messages are very useful for future code debugging. They also provide documentation for quality assurance testers and other developers.
+For instance, to add an error message to `self.assertEqual(double(8), 16)`, change your code to match this:
+
+```python
+   def test_double(self):
+      self.assertEqual(double(8), 16, "the function should return two times the number provided")
+```
+
+###### Output
+
+```console
+>>> python3 -m unittest main.py
+.F.
+======================================================================
+FAIL: test_double (__main__.TestMathFunctions)
+----------------------------------------------------------------------
+Traceback (most recent call last):
+  File "main.py", line 8, in test_double
+    self.assertEqual(double(8), 16, "the function should return two times the number provided")
+AssertionError: 32 != 16 : the function should return two times the number provided
+
+----------------------------------------------------------------------
+Ran 3 tests in 0.001s
+
+FAILED (failures=1)
+```
+
+As demonstrated above, the error message we added describes the `AssertionError` better. This has made it more understandable.
+
+##### More assert methods
+So far, we've seen the `assertEqual(a, b)` method, the `assertNotEqual(a, b)` method, and the `assertRaises(x)` method. Apart from these, the unittest module provides extra assert methods which include:
+
+1. `assertTrue(x)` checks that a boolean value `x` equals `True`.
+2. `assertFalse(x)` checks that a boolean value `x` equals `False`.
+3. `assertIn(a,b)` checks that a value `a` exists in a list `b`. For instance, this check should pass when `a=2` and `b=[1,2,3]`.
+4. `asserNottIn(a,b)` checks that a value `a` does not exist in a list `b`. For instance, this check should fail when `a=2` and `b=[1,2,3]` since `2` exists in the list `b`.
+
+Once you are comfortable working with these, you can check more assert methods [here](https://docs.python.org/3/library/unittest.html#assert-methods).
+
+### Benefits of Unit Testing
+Unit testing is an essential step of the software development life cycle and provides benefits such as:
+
+1. Unit tests provide an efficient and automated way to discover bugs.
+2. Unit testing ensures that a unit works properly after refactoring (regression testing).
+3. Detecting bugs during the early phases of development reduces project costs.
+4. The unit tests can help new developers familiarize themselves with the working of the units.
+5. Unit tests provide code quality assurance since they show that it is functioning correctly.
```

Comment:
1. Efficient and automated way to discover bugs.
2. Ensures that a unit works properly after refactoring (regression testing).
3. Detecting bugs during the early phases of development reduces project costs.
4. It can help new developers familiarize themselves with the working of the units.
5. It provides code quality assurance since they show it is functioning correctly.